src/Raven.Server/Documents/Indexes/Persistence/Corax/QueryOptimizer/CoraxBooleanItem.cs

```diff
@@ -49,63 +48,121 @@ private CoraxBooleanItem(IndexSearcher searcher, Index index, FieldMetadata fiel
+        return field.HasBoost 
+            ? cbi.Materialize(ref streamingOptimization) 
+            : cbi;
     }
-    
-    private CoraxBooleanItem(IndexSearcher searcher, Index index, FieldMetadata field, object term1, object term2, UnaryMatchOperation operation, UnaryMatchOperation left, UnaryMatchOperation right) : this(searcher, index, field, term1, operation)
+
+    public static IQueryMatch BuildBetween(IndexSearcher indexSearcher, Index index, FieldMetadata field, object leftValue, object rightValue,
+        UnaryMatchOperation leftOperator, UnaryMatchOperation rightOperator, ref CoraxQueryBuilder.StreamingOptimization streamingOptimization)
     {
-        //Between handler
-        
-        if (_isTime) //found time at `Term1`, lets check if second item also contains time
+        var leftIsUnbounded = leftValue is null or Constants.Documents.Querying.Terms.LeftNullValueOfBetweenQuery;
+        var rightIsUnbounded = rightValue is null or Constants.Documents.Querying.Terms.RightNullValueOfBetweenQuery;
+
+        switch (IsLeftUnbounded: leftIsUnbounded, IsRightUnbounded: rightIsUnbounded)
         {
-            if (term2 != null && index.IndexFieldsPersistence.HasTimeValues(field.FieldName.ToString()) && QueryBuilderHelper.TryGetTime(index, term2, out var ticks))
+            case (IsLeftUnbounded: true, IsRightUnbounded: true):
             {
-                Term2 = ticks;
+                Debug.Assert(streamingOptimization.OptimizationIsPossible == false);
+                var existsQuery = indexSearcher.ExistsQuery(field, streamingEnabled: false, forward: true);
+                
+                // matching lucene results, nulls included
+                return indexSearcher.IncludeNullMatch(field, existsQuery, forward: false);
             }
-            else  //not found, lets revert Term1 
+            
+            case (IsLeftUnbounded: true, IsRightUnbounded: false):
             {
-                Term = term1;
-                Term2 = term2;
-                _isTime = false;
+                Debug.Assert(streamingOptimization.OptimizationIsPossible == false);
+                // between null and Value => (oo, x)
+                return Build(indexSearcher, index, field, rightValue, rightOperator, ref streamingOptimization);
             }
-        }
-        else
-        {
-            Term2 = term2;
-        }
 
-        BetweenRight = right;
-        BetweenLeft = left;
+            case (IsLeftUnbounded: false, IsRightUnbounded: true):
+            {
+                Debug.Assert(streamingOptimization.OptimizationIsPossible == false);
+                // between Value and null => (x, oo)
+                var query = Build(indexSearcher, index, field, leftValue, leftOperator, ref streamingOptimization, out bool isNumerical);
+
+                // For numerical queries we will include the null values for backward compatibility.
+                if (isNumerical == false)
+                    return query;
```

Comment:
Why aren't we including that for everything?

Reply:
Wouldn't it be better to state that null is ordered first or last (whatever would be consistent) and that is it?
That means that we have internally consistent behavior for Corax

Lucene won't be with us forever

Reply:
I think we can do it consistently to the numeric ranges behavior.  I pushed the change.